### PR TITLE
ON-44919 # Added CONCURRENT approval steps to SDK

### PR DESCRIPTION
**Pull request by Zaxist**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [x] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

### CODE REVIEW

File: src/lib/forms-schema/index.ts [hunk 1]
```diff
@@ -523,26 +534,30 @@ const formSchema = Joi.object().keys({
...
+          ...approvalStepCommonProps,
+        }).required(),
+        Joi.object({
+          type: Joi.string().valid('CONCURRENT').required(),
+          nodes: Joi.array()
+            .items(
+              Joi.object({
+                label: Joi.string().required(),
+                group: Joi.string().required(),
+                approvalFormId: Joi.number(),
```
Comment (jdawg093) [inline src/lib/forms-schema/index.ts:553]: Should we split this out into its own common schema in case we have changes in the future that we want for steps?

File: package.json [hunk 2]
```diff
@@ -21,7 +21,7 @@
     "@aws-sdk/client-s3": "^3.637.0",
     "@microsoft/eslint-plugin-sdl": "^0.2.2",
     "@oneblink/release-cli": "^3.2.1",
-    "@oneblink/types": "github:oneblink/types",
+    "@oneblink/types": "github:oneblink/types#ON-44770",
```
Comment (divporter) [inline package.json:24]: fix plz

File: src/lib/forms-validation/index.ts [hunk 3]
```diff
@@ -272,6 +272,25 @@ function validateWithFormSchema(form?: unknown):
       }
     }
 
+    const labels: string[] = []
```
Comment (divporter) [inline src/lib/forms-validation/index.ts:275]: I don't see any code to push items into this array, but perhaps I'm an idiot

File: tests/forms-schema/forms-schema.test.ts [hunk 4]
```diff
@@ -8004,3 +8004,109 @@ describe('Slug', () => {
...
+            label: 'Label 2',
+            approvalFormId: 1,
+            conditionalPredicates: [],
+            isConditional: false,
+            requiresAllConditionalPredicates: false,
+          },
+        ],
+      })
+    }).toThrow('"approvalSteps[0]" does not match any of the allowed types')
+  })
```
Comment (divporter) [inline tests/forms-schema/forms-schema.test.ts:8111]: add a test for non-unique labels

File: src/lib/forms-validation/index.ts (hunk 3 above)
Comment (Zaxist) [reply src/lib/forms-validation/index.ts:275]: Harr yep.

File: tests/forms-schema/forms-schema.test.ts (hunk 4 above)
Comment (Zaxist) [reply tests/forms-schema/forms-schema.test.ts:8111]: Nearly did too, and went the lazy route. How embarrassing.
